crates/matrix-sdk-base/src/store/mod.rs

```diff
@@ -319,6 +270,175 @@ impl Deref for Store {
     }
 }
 
+/// An observable map.
+///
+/// This is an “observable map” naive implementation. Just like regular hashmap,
+/// we have a redirection from a key to a position, and from a position to
+/// a value. The (key, position) tuples are stored in an [`HashMap`]. The
+/// (position, value) tuples are stored in an [`ObservableVector`]. The (key,
+/// position) tuple is only provided for fast _reading_ implementations, like
+/// `Self::get` and `Self::get_or_create`. The (position, value) tuples are
+/// observable, this is what interests us the most here.
+///
+/// Why not implementing a new `ObservableMap` type in `eyeball-im` instead of
+/// this custom implementation? Because we want to continue providing
+/// `VectorDiff` when observing the changes, so that the rest of the API in the
+/// Matrix Rust SDK aren't broken. Indeed, an `ObservableMap` must produce
+/// `MapDiff`, which would be quite different.
+/// Plus, we would like to re-use all our existing code, test, stream adapters
+/// and so on.
+///
+/// This is a trade-off. And this implementation is simple enough for the
+/// moment, and basically does the job.
+#[derive(Debug)]
+struct ObservableMap<K, V>
```

Comment:
Ah thanks, it's a better name. I like it when you address my review comments before I've sent the review.

Reply:
We are connected.